tests/assembly/s390x-backchain-toggle.rs

```diff
@@ -0,0 +1,41 @@
+//@ revisions: enable-backchain disable-backchain
+//@ assembly-output: emit-asm
+//@ compile-flags: -O --crate-type=lib --target=s390x-unknown-linux-gnu
+//@ needs-llvm-components: systemz
+//@[enable-backchain] compile-flags: -Ctarget-feature=+backchain
+//@[disable-backchain] compile-flags: -Ctarget-feature=-backchain
+#![feature(no_core, lang_items)]
+#![no_std]
+#![no_core]
+
+#[lang = "sized"]
+trait Sized {}
+
+extern "C" {
+    fn extern_func();
+}
+
+// CHECK-LABEL: test_backchain
+#[no_mangle]
+extern "C" fn test_backchain() -> i32 {
+    // enable-backchain: lgr [[REG1:.*]], %r15
+    // enable-backchain-NEXT: aghi %r15, -160
+    // enable-backchain: stg [[REG1]], 0(%r15)
+    // disable-backchain: aghi %r15, -160
+    // disable-backchain-NOT: stg %r{{.*}}, 0(%r15)
+    unsafe {
+        extern_func();
+    }
+    // enable-backchain-NEXT: brasl %r{{.*}}, extern_func@PLT
+    // enable-backchain-NEXT: lghi %r2, 1
+    // disable-backchain: lghi %r2, 0
```

Comment:
Is `lghi %r2` setting the return value here?

(I can't read assembly code, I am afraid. I would have written this test differently but I guess the stuff above, with `extern_func`, tests whether "backchain" is passed to LLVM? So this test covers more than what I could have covered.)

Reply:
> Is `lghi %r2` setting the return value here?

Yes, `lghi` on s390x is "Load Halfword Immediate". `%r2` register is the first return argument register on s390x when running a Linux-based system in Z-mode (see https://github.com/IBM/s390x-abi/releases/tag/v1.6.1, the addresses are 64-bit).

> (I can't read assembly code, I am afraid. I would have written this test differently but I guess the stuff above, with extern_func, tests whether "backchain" is passed to LLVM? So this test covers more than what I could have covered.)

The test is an assembly scanning test because different LLVM versions may interpret attributes differently. Scanning the resulting assembly is a more robust way to detect if the "blockchain" option is passed down correctly.